synedrion/src/cggmp21/sigma/aff_g_star.rs

```diff
@@ -0,0 +1,338 @@
+    /// Point $X = g^x$, where $g$ is the curve generator.
+    pub cap_x: &'a Point,
+}
+
+struct AffGStarProofEphemeral<P: SchemeParams> {
+    alpha: SecretSigned<<P::Paillier as PaillierParams>::Uint>,
+    beta: SecretSigned<<P::Paillier as PaillierParams>::Uint>,
+    r: Randomizer<P::Paillier>,
+    s: Randomizer<P::Paillier>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct AffGStarProofCommitment<P: SchemeParams> {
+    cap_a: CiphertextWire<P::Paillier>,
+    cap_r: Point,
+    cap_b: CiphertextWire<P::Paillier>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct AffGStarProofElement<P: SchemeParams> {
+    z: PublicSigned<<P::Paillier as PaillierParams>::Uint>,
+    z_prime: PublicSigned<<P::Paillier as PaillierParams>::Uint>,
+    w: MaskedRandomizer<P::Paillier>,
+    lambda: MaskedRandomizer<P::Paillier>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(bound(serialize = "
+    AffGStarProofCommitment<P>: Serialize,
+    AffGStarProofElement<P>: Serialize,
+"))]
+#[serde(bound(deserialize = "
+    AffGStarProofCommitment<P>: for<'x> Deserialize<'x>,
+    AffGStarProofElement<P>: for<'x> Deserialize<'x>,
+"))]
+pub(crate) struct AffGStarProof<P: SchemeParams> {
+    e: BitVec,
+    commitments: Box<[AffGStarProofCommitment<P>]>,
+    elements: Box<[AffGStarProofElement<P>]>,
+}
+
+impl<P: SchemeParams> AffGStarProof<P> {
+    pub fn new(
+        rng: &mut impl CryptoRngCore,
+        secret: AffGStarSecretInputs<'_, P>,
+        public: AffGStarPublicInputs<'_, P>,
+        aux: &impl Hashable,
+    ) -> Self {
+        secret.x.assert_exponent_range(P::L_BOUND);
+        secret.y.assert_exponent_range(P::LP_BOUND);
+        assert!(public.cap_c.public_key() == public.pk0);
+        assert!(public.cap_d.public_key() == public.pk0);
+        assert!(public.cap_y.public_key() == public.pk1);
+
+        let (ephemerals, commitments): (Vec<_>, Vec<_>) = (0..P::SECURITY_PARAMETER)
+            .map(|_| {
+                let alpha = SecretSigned::random_in_exponent_range(rng, P::L_BOUND + P::EPS_BOUND);
+                let beta = SecretSigned::random_in_exponent_range(rng, P::LP_BOUND + P::EPS_BOUND);
+                let r = Randomizer::random(rng, public.pk0);
+                let s = Randomizer::random(rng, public.pk1);
+
+                let cap_a = (public.cap_c * &alpha + Ciphertext::new_with_randomizer(public.pk0, &beta, &r)).to_wire();
+                let cap_r = secret_scalar_from_signed::<P>(&alpha).mul_by_generator();
+                let cap_b = Ciphertext::new_with_randomizer(public.pk1, &beta, &s).to_wire();
+
+                let ephemeral = AffGStarProofEphemeral::<P> { alpha, beta, r, s };
+                let commitment = AffGStarProofCommitment { cap_a, cap_r, cap_b };
+
+                (ephemeral, commitment)
+            })
+            .unzip();
+
+        let mut reader = XofHasher::new_with_dst(HASH_TAG)
+            // commitments
+            .chain(&commitments)
```

Comment:
This is the Shamir transform right? How can I double check (against the paper?) that we actually have all the data that we need baked into the reader? Does the order of the params matter here you reckon?

Reply:
The background to my question is vulnerabilities like [Frozen Heart](https://blog.trailofbits.com/2022/04/13/part-1-coordinated-disclosure-of-vulnerabilities-affecting-girault-bulletproofs-and-plonk/) ([paper](https://eprint.iacr.org/2023/691.pdf)). 
Applied here, my question is: how can we make sure that a) the code in this spot and elsewhere is correct (I guess the answer is: "trust the paper and make sure we're doing exactly what they write") and b) can we come up with a way to machine-check this hashing to ensure it stays correct?

Reply:
If you mean correct in general, then yes, just try to faithfully reproduce the paper and blame everything on them :) Well, and apply some common sense, because the paper does have plenty of typos.

But there may be some specific things we can enforce by typing and code structure. Specifically, we can make it a little harder to make a mistake in making a challenge by closing #110 and #182